GraphQL Client: Continue advice on using or bypassing Apollo's cache

courier-new opened the pull request:
Follow-up to #22166 that adds some additional advice/recommendations around when (and when not!) to use Apollo's cache. Basically integrated and expanded upon the comment I left on that PR. 🙂

umpox reviewed (approved):
Great writing 👍

5h1rU commented:
I think we are missing the case when you show a message like "3 new repositories were added. Do you want to add them?". You will have the data optimistically but add this to the UI is up to the user.
Also, for paginated data, it's not necessarily a problem. I'm not a big fan of encouraging spinners as a solution when this already produces a significant cls in most cases. 

Do we need to add this scenario as part of our design process? @sourcegraph/design

courier-new commented:
> I think we are missing the case when you show a message like "3 new repositories were added. Do you want to add them?". You will have the data optimistically but add this to the UI is up to the user.

That's a great suggestion! I'll add that to the "strategies" at the end of this section.

> Also, for paginated data, it's not necessarily a problem. I'm not a big fan of encouraging spinners as a solution when this already produces a significant cls in most cases.

Definitely agree, I personally think showing the stale data **with** a loading spinner strategy or the user prompt to add the new data like you suggested work best for paginated data. I'd love to make this clearer in the recommendations, though. Would you say I'm over-encouraging the use of spinners a bit? Let me try to reword a couple places.

> Do we need to add this scenario as part of our design process?

Good question. Ideally probably yes, but I'd say most of the time it's probably fine to leave it up to the developer's judgment, in the same way that most developers probably decide for themselves how to update the UI over the course of a mutation running (any loading state, optimistic updates, and the final response state/errors) and get design to sign off on the usability of it.

courier-new commented:
@5h1rU I've made some edits in the hopes of addressing your concerns, but I'll go ahead and merge this for now and we can continue iterating upon and improving these as we roll out Apollo across the codebase! 🙂

5h1rU commented:
> @5h1rU I've made some edits in the hopes of addressing your concerns, but I'll go ahead and merge this for now and we can continue iterating upon and improving these as we roll out Apollo across the codebase! 🙂

@courier-new  Perfect, thank you so much for this writing. 🎉
